zkchat-example/src/main.rs

```diff
@@ -0,0 +1,464 @@
+        parallelism: 3,
+        record_store: (),
+        peer_store: peer_store,
+        local_peer_id: my_peer_id.clone(),
+        timeout: Duration::from_secs(2),
+    };
+
+    let kad_ctl_proto = libp2p_kad::KademliaControllerPrototype::new(kad_config);
+    let kad_upgrade = libp2p_kad::KademliaUpgrade::from_prototype(&kad_ctl_proto);
+
+    let upgrade = ConnectionUpgrader {
+        kad: kad_upgrade,
+        identify: libp2p_identify::IdentifyProtocolConfig,
+        floodsub: floodsub_upgrade.clone(),
+    };
+
+    let listened_addrs = Arc::new(RwLock::new(vec![]));
+    let listened_addrs_clone = listened_addrs.clone();
+
+    let my_peer_id_clone = my_peer_id.clone();
+    let (swarm_controller, swarm_future) = libp2p_swarm::swarm(
+        transport,
+        upgrade,
+        move |upgrade, client_addr| match upgrade {
+            FinalUpgrade::Kad(kad) => Box::new(kad) as Box<_>,
+            FinalUpgrade::FloodSub(future) => Box::new(future) as Box<_>,
+            FinalUpgrade::Identify(IdentifyOutput::Sender { sender, .. }) => sender.send(
+                IdentifyInfo {
+                    public_key: my_peer_id_clone.clone().into_bytes(),
+                    protocol_version: "chat/1.0.0".to_owned(),
+                    agent_version: "rust-libp2p/1.0.0".to_owned(),
+                    listen_addrs: listened_addrs_clone.read().unwrap().to_vec(),
+                    protocols: vec![
+                        "/ipfs/kad/1.0.0".to_owned(),
+                        "/ipfs/id/1.0.0".to_owned(),
+                        "/floodsub/1.0.0".to_owned(),
+                    ],
+                },
+                &client_addr,
+            ),
+            FinalUpgrade::Identify(IdentifyOutput::RemoteInfo { .. }) => {
+                unreachable!("We are never dialing with the identify protocol")
+            }
+        },
+    );
+
+    // Listening on the address passed as program argument, or a default one.
+    if cfg!(not(target_os = "emscripten")) {
+        let parsed: Multiaddr = if is_bootstrap {
+            "/ip4/0.0.0.0/tcp/10101/ws".parse().unwrap()
+        } else {
+            let addr = env::args()
+                .skip(1)
+                .next()
+                .unwrap_or("/ip4/0.0.0.0/tcp/0/ws".to_owned());
+            addr.parse()
+                .expect("failed to parse multiaddress to listen to")
+        };
+
+        let actual_addr = swarm_controller
+            .listen_on(parsed.clone())
+            .expect("failed to listen to multiaddress");
+        info!("Now listening on {}", actual_addr);
+        listened_addrs.write().unwrap().push(actual_addr); // TODO: wrong, requires nat_traversal
+    }
+
+    let (kad_controller, kad_init) = kad_ctl_proto.start(swarm_controller.clone());
+
+    let topic = libp2p_floodsub::TopicBuilder::new("chat").build();
+
+    let floodsub_ctl =
+        libp2p_floodsub::FloodSubController::new(&floodsub_upgrade, swarm_controller.clone());
```

Comment:
Why do we need swarm controller for floodsub? can we omit it in FloodSubController struct?

Reply:
True! That's a mistake.
Since this concerns a code already merged, could you open an issue instead?

Reply:
@tomaka created a PR: https://github.com/libp2p/rust-libp2p/pull/160 , please review, thanks.